File: languages/csharp/SecureMemory/SecureMemory/Secret.cs
using System;

namespace GoDaddy.Asherah.SecureMemory
{
    public abstract class Secret : IDisposable
    {
        public abstract TResult WithSecretBytes<TResult>(Func<byte[], TResult> funcWithSecret);

        public abstract TResult WithSecretUtf8Chars<TResult>(Func<char[], TResult> funcWithSecret);

        public virtual void WithSecretBytes(Action<byte[]> actionWithSecret)
        {
            WithSecretBytes(bytes =>
            {
                actionWithSecret(bytes);
                return true;
            });
        }

        public virtual void WithSecretUtf8Chars(Action<char[]> actionWithSecret)
        {
            WithSecretUtf8Chars(chars =>
            {
                actionWithSecret(chars);
                return true;
            });
        }

        public abstract Secret CopySecret();

        public abstract void Close();

        public abstract void Dispose();
    }
}

File: java/secure-memory/src/main/java/com/godaddy/asherah/securememory/Secret.java
package com.godaddy.asherah.securememory;

import java.util.function.Consumer;
import java.util.function.Function;

public interface Secret {
  <T> T withSecretBytes(Function<byte[], T> function);

  <T> T withSecretUtf8Chars(Function<char[], T> function);

  default void withSecretBytes(Consumer<byte[]> consumer) {
    this.withSecretBytes(bytes -> {
      consumer.accept(bytes);
      return null;
    });
  }

  default void withSecretUtf8Chars(Consumer<char[]> consumer) {
    this.withSecretUtf8Chars(chars -> {
      consumer.accept(chars);
      return null;
    });
  }

  Secret copySecret();

  void close();
}
